presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java

```diff
@@ -692,34 +689,101 @@ protected ExtractionResult visitNullLiteral(NullLiteral node, Boolean complement
 
         Type leftType = expressionTypes.get(comparison.getLeft());
         Type rightType = expressionTypes.get(comparison.getRight());
-
         Preconditions.checkArgument(leftType.equals(rightType), "left and right type do not match in comparison expression (%s)", comparison);
 
-        if (left instanceof SymbolReference && !(right instanceof Expression)) {
-            return Optional.of(new NormalizedSimpleComparison((SymbolReference) left, comparison.getType(), new NullableValue(rightType, right)));
+        if (left instanceof Expression && right instanceof Expression) {
+            return Optional.empty();
+        }
+
+        if (!(left instanceof Expression) && !(right instanceof Expression)) {
+            return Optional.empty();
+        }
+
+        Expression symbolExpression;
+        ComparisonExpressionType comparisonType;
+        NullableValue value;
+
+        if (left instanceof Expression) {
+            symbolExpression = comparison.getLeft();
+            comparisonType = comparison.getType();
+            value = new NullableValue(rightType, right);
+        }
+        else {
+            symbolExpression = comparison.getRight();
+            comparisonType = comparison.getType().flip();
+            value = new NullableValue(leftType, left);
+        }
+
+        if (!checkSymbolExpressionValid(metadata, expressionTypes, symbolExpression)) {
+            return Optional.empty();
         }
-        if (right instanceof SymbolReference && !(left instanceof Expression)) {
-            return Optional.of(new NormalizedSimpleComparison((SymbolReference) right, comparison.getType().flip(), new NullableValue(leftType, left)));
+
+        return Optional.of(new NormalizedSimpleComparison(symbolExpression, comparisonType, value));
+    }
+
+    private static Type typeOf(Expression expression, Session session, Metadata metadata, Map<Symbol, Type> types)
+    {
+        IdentityHashMap<Expression, Type> expressionTypes = ExpressionAnalyzer.getExpressionTypes(session, metadata, new SqlParser(), types, expression, emptyList() /* parameters already replaced */);
+        return expressionTypes.get(expression);
+    }
+
+    private static boolean checkSymbolExpressionValid(Metadata metadata, IdentityHashMap<Expression, Type> expressionTypes, Expression symbolExpression)
+    {
+        while (true) {
+            if (symbolExpression instanceof SymbolReference) {
+                return true;
+            }
+            else if (symbolExpression instanceof Cast) {
+                if (!isCoercion(metadata, expressionTypes, (Cast) symbolExpression)) {
```

Comment:
After thinking about this in more depth, I don't think we can handle this operation in this fashion, even with only unrolling implicit coercions.

For example, I think you mentioned this example before, but with the case
```
CAST(my_bigint AS DOUBLE) = 1.1
```
This is a valid implicit coercion, but it is not equivalent to my_bigint = 1.

We may be able to make this work if we can succinctly find a way to define how implicit coercions operate. For example, we may need to handle the different types of comparisons in different ways (sort of like how rewriteComparisonExpression does that).